### PR TITLE
[FLINK-30113] Implement state compression for broadcast and regular operator states

### PR DESCRIPTION
**Pull request by echauchot**
## What is the purpose of the change

Implement state compression for broadcast and regular operator states. Only the states themselves are compressed not the metadata.

## Brief change log

  -  Introduce FSData streams wrappers that add compression
  - Use these wrappers while writing state (in the default operator snapshot strategy) and while reading state (in the operator state restore operation)

## Verifying this change

This change is already covered by existing tests, such as windowing ITCase tests ... in the flink-runtime module

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: yes checkpointing
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? javadocs

R: @dawidwys

### CODE REVIEW

**Comment by dawidwys** [comment]
Thanks for the PR. I'll review it over the next few days.

**Comment by MartijnVisser** [comment]
@dawidwys Could you review this PR?

**Review by dawidwys — CHANGES_REQUESTED** [review]
Hey @echauchot 
Thanks for working on the issue.

Correct me if I am wrong, but in the current state you always apply the compression for the operator state. 

First of all, imo, this needs to be configurable. I'd say it can be done with the same flag as for the keyed state backend (`execution.checkpointing.snapshot-compression`)

Moreover, right now it causes previously written checkpoints to be unreadable. (have you tried restoring from an old checkpoint? If not, could you add such a test?) as it always assumes the stream is compressed. The info if the state is compressed or not needs to be persisted somewhere. I don't have a good idea for that now. In the keyed state case it is done via `KeyedBackendSerializationProxy#read/write`.

Now I realise this might require changes to the checkpoint format, unless we figure out a smart place to put it, so we might need to create a small FLIP for that.

**Comment by echauchot** [comment]
> Hey @echauchot Thanks for working on the issue.
> 
> Correct me if I am wrong, but in the current state you always apply the compression for the operator state.
> 

No, see `DefaultOperatorStateBackendBuilder` I call `AbstractStateBackend.getCompressionDecorator` that reads `execution.checkpointing.snapshot-compression` conf parameter.

> First of all, imo, this needs to be configurable. I'd say it can be done with the same flag as for the keyed state backend (`execution.checkpointing.snapshot-compression`)

See above, I've found we could reuse `execution.checkpointing.snapshot-compression` conf parameter. Tell me if you want it to be a different conf parameter. 

 
> Moreover, right now it causes previously written checkpoints to be unreadable. (have you tried restoring from an old checkpoint? If not, could you add such a test?) as it always assumes the stream is compressed. 
> The info if the state is compressed or not needs to be persisted somewhere. I don't have a good idea for that now. In the keyed state case it is done via `KeyedBackendSerializationProxy#read/write`.

Well, if the user disables compression to read uncompressed snapshots, then it works.
Though, I have not tested to restore old snapshots. I have just tested the state write/read roundtrip via existing tests. I'll add such a test

Yes I saw that with key state a boolean is written first when the state is written. I can do similar thing for operator state.

> 
> Now I realise this might require changes to the checkpoint format, unless we figure out a smart place to put it, so we might need to create a small FLIP for that.

Ok, adding the boolean will require to change the snapshot format indeed. This is why I have not added this boolean to avoid the format change. But, indeed, if a snapshot was compressed and the user then disables compression he could not read the previous compressed snapshot with my code. What about simply adding a compression boolean  to operator state meta ? Need a FLIP for this change ?

**Comment by dawidwys** [comment]
> No, see DefaultOperatorStateBackendBuilder I call AbstractStateBackend.getCompressionDecorator that reads execution.checkpointing.snapshot-compression conf parameter.

Ah, sorry missed that :(

> Ok, adding the boolean will require to change the snapshot format indeed. This is why I have not added this boolean to avoid the format change. But, indeed, if a snapshot was compressed and the user then disables compression he could not read the previous compressed snapshot with my code. What about simply adding a compression boolean to operator state meta ? As a backward incompatible change, it requires a FLIP even if simple change, right ?

Yes, I do think it requires a FLIP as it is a substantial change, but I believe this change is unavoidable if we want to add operator state compression.

**Comment by echauchot** [comment]
> > No, see DefaultOperatorStateBackendBuilder I call AbstractStateBackend.getCompressionDecorator that reads execution.checkpointing.snapshot-compression conf parameter.
> 
> Ah, sorry missed that :( That part looks good then.
> 
> > Ok, adding the boolean will require to change the snapshot format indeed. This is why I have not added this boolean to avoid the format change. But, indeed, if a snapshot was compressed and the user then disables compression he could not read the previous compressed snapshot with my code. What about simply adding a compression boolean to operator state meta ? As a backward incompatible change, it requires a FLIP even if simple change, right ?
> 
> Yes, I do think it requires a FLIP as it is a substantial change, but I believe this change is unavoidable if we want to add operator state compression.

Ok, I'll write the FLIP and ping you on it for review, then when we agree on the architecture, I'll write the code and commit on this PR.

**Comment by echauchot** [comment]
> Ok, I'll write the FLIP and ping you on it for review, then when we agree on the architecture, I'll write the code and commit on this PR.

@dawidwys I don't have the rights to add a FLIP to Flink confluence space. Can you give me the rights ?

**Comment by MartijnVisser** [comment]
> Can you give me the rights ?

@echauchot You should have the rights now

**Comment by echauchot** [comment]
> > Can you give me the rights ?
> 
> @echauchot You should have the rights now

Thanks @MartijnVisser !

**Comment by echauchot** [comment]
> > Ok, I'll write the FLIP and ping you on it for review, then when we agree on the architecture, I'll write the code and commit on this PR.
> 
> @dawidwys I don't have the rights to add a FLIP to Flink confluence space. Can you give me the rights ?

@dawidwys [here](https://cwiki.apache.org/confluence/display/FLINK/FLIP-289+Operator+state+compression) is the flip, cn you please review it ?

**Comment by echauchot** [comment]
The vote on the FLIP has passed, but I'm going on vacation tonight for a week. I'll update this PR to reflect the FLIP when I get back

**Comment by dawidwys** [comment]
Absolutely @echauchot ! Thanks for the work so far!

**Comment by echauchot** [comment]
@dawidwys sorry for the delay, I had to finish [this other PR](https://github.com/apache/flink-connector-cassandra/pull/3) after my vacations. Resuming the work on this PR ...

**Comment by echauchot** [comment]
@flinkbot run azure

**Comment by echauchot** [comment]
@flinkbot run azure

**Comment by echauchot** [comment]
Hi @dawidwys, my code is failing with versioning. I incremented the proxy version (as the metadata serde code has changed) and the snapshot version (as I added a boolean for compression). Can you explain how the proxy/snapshot versioning works because I see nothing in the docs.

**Comment by echauchot** [comment]
@flinkbot run azure

**Comment by echauchot** [comment]
@flinkbot run azure

**Comment by echauchot** [comment]
> What you did for versioning looks sane. I think the problem is with line
> 
> https://github.com/apache/flink/blob/a4a11a389609deb5d1de139115cedc35b675b366/flink-runtime/src/main/java/org/apache/flink/runtime/state/metainfo/StateMetaInfoSnapshotReadersWriters.java#L68
> 
> which has not been implemented in a future proof way. The intent there was to throw an exception for `readVersion < 6`, not less than `CURRENT_STATE_META_INFO_SNAPSHOT_VERSION`. If we change the condition to `readVersion < 6` it should work.

Hi @dawidwys, thanks for the prompt answer. I suspected something not future proof, thanks for the confirmation. It is now fixed. Tests pass but the CI fails in preparation of the E2E tests (since some days) with a http 404 error.

PTAL.

**Review by dawidwys — APPROVED** [review]
LGTM, let's wait until e2e tests are fixed though before we merge it.

**Comment by MartijnVisser** [comment]
> LGTM, let's wait until e2e tests are fixed though before we merge it.

These E2E tests have been fixed for some time already, you probably need to rebase :) See https://issues.apache.org/jira/browse/FLINK-30972

**Comment by echauchot** [comment]
> > LGTM, let's wait until e2e tests are fixed though before we merge it.
> 
> These E2E tests have been fixed for some time already, you probably need to rebase :) See https://issues.apache.org/jira/browse/FLINK-30972

Thanks @MartijnVisser, I missed this ticket. I just rebased. Let's wait for the green lights.

**Comment by echauchot** [comment]
@flinkbot run azure

**Comment by echauchot** [comment]
@dawidwys thanks for reviewing/merging.

**Pull request merged**
